proof-systems/src/darlin/pcd/mod.rs

```diff
@@ -0,0 +1,232 @@
+        &self,
+        vk:         &Self::PCDVerifierKey,
+        rng:        &mut R,
+    ) -> Result<bool, PCDError>
+    {
+        let acc = self.succinct_verify(vk)?;
+        self.hard_verify::<R>(acc, vk, rng)
+    }
+}
+
+#[derive(Derivative)]
+#[derivative(Clone(bound = ""))]
+/// Achieve polymorphism for PCD via an enumerable. This provides nice APIs for
+/// the proof aggregation implementation and testing.
+pub enum GeneralPCD<'a, G1: AffineCurve, G2: AffineCurve, D: Digest> {
+    SimpleMarlin(SimpleMarlinPCD<'a, G1, D>),
+    FinalDarlin(FinalDarlinPCD<'a, G1, G2, D>)
+}
+
+// Testing functions
+impl<'a, G1, G2, D> GeneralPCD<'a, G1, G2, D>
+    where
+        G1: AffineCurve<BaseField = <G2 as AffineCurve>::ScalarField> + ToConstraintField<<G2 as AffineCurve>::ScalarField>,
+        G2: AffineCurve<BaseField = <G1 as AffineCurve>::ScalarField> + ToConstraintField<<G1 as AffineCurve>::ScalarField>,
+        D: Digest,
+{
+    pub fn randomize_usr_ins<R: RngCore>(
+        &mut self,
+        rng: &mut R
+    )
+    {
+        match self {
+            Self::SimpleMarlin(simple_marlin) => {
+                // No sys ins (for now) for SimpleMarlin, so modify the usr_ins instead
+                let ins_len = simple_marlin.usr_ins.len();
+                simple_marlin.usr_ins = vec![G1::ScalarField::rand(rng); ins_len];
+            },
+            Self::FinalDarlin(final_darlin) => {
+                let ins_len = final_darlin.usr_ins.len();
+                final_darlin.usr_ins = vec![G1::ScalarField::rand(rng); ins_len];
+            }
+        }
+    }
+
+    pub fn randomize_sys_ins<R: RngCore>(
+        &mut self,
+        ck_g1: &DLogCommitterKey<G1>,
+        ck_g2: &DLogCommitterKey<G2>,
+        rng: &mut R
+    )
+    {
+        match self {
+            Self::SimpleMarlin(simple_marlin) => {
+                // No sys ins (for now) for SimpleMarlin, so modify the usr_ins instead
+                let ins_len = simple_marlin.usr_ins.len();
+                simple_marlin.usr_ins = vec![G1::ScalarField::rand(rng); ins_len];
+            },
+            Self::FinalDarlin(final_darlin) => {
+                final_darlin.final_darlin_proof.deferred = FinalDarlinDeferredData::<G1, G2>::generate_random::<R, D>(
+                    rng, ck_g1, ck_g2
+                );
+            }
+        }
+    }
+}
+
+impl<'a, G1, G2, D> PCD for GeneralPCD<'a, G1, G2, D>
+where
+    G1: AffineCurve<BaseField = <G2 as AffineCurve>::ScalarField> + ToConstraintField<<G2 as AffineCurve>::ScalarField>,
+    G2: AffineCurve<BaseField = <G1 as AffineCurve>::ScalarField> + ToConstraintField<<G1 as AffineCurve>::ScalarField>,
+    D: Digest + 'a,
+{
+    type PCDAccumulator = DualDLogItemAccumulator<'a, G1, G2, D>;
+    // a work-around: we use throughout aFinalDarlinPCDVerifierKey for both 
+    // final Marlin and simple Marlin PCDs, as it contains both dlog vks.
+    type PCDVerifierKey = FinalDarlinPCDVerifierKey<'a, G1, G2, D>;
```

Comment:
We should double-check if there is a more elegant solution for that.

Reply:
Created a new type alias for FinalDarlinPCDVerifierKey